src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -194,20 +194,91 @@ public override int GetHashCode()
     /// </summary>
     public class CallGraph
     {
-        // TODO:
-        // This is the method that should be invoked to verify cycles of interest,
-        // i.e. where each callable in the cycle is type parametrized.
-        // It should probably generate diagnostics; I'll add the doc comment once its use is fully defined.
-        internal static bool VerifyCycle(CallGraphNode rootNode, params CallGraphEdge[] edges)
+        /// <summary>
+        /// Given a cycle of call graph edges, determines if the cycle is valid.
+        /// Invalid cycles are those that cause type parameters to be mapped to
+        /// other type parameters of the same callable (constricting resolutions)
+        /// or to a type containing a nested reference to the same type parameter,
+        /// i.e Foo.A -> Foo.A[].
```

Comment:
FYI, Haskell calls this an infinite type:

```
Prelude> f x = x : [f x]

<interactive>:12:1: error:
    * Occurs check: cannot construct the infinite type: t ~ [t]
      Expected type: t -> t
        Actual type: t -> [t]
    * Relevant bindings include
        f :: t -> t (bound at <interactive>:12:1)
```

Reply:
Looks like this hassle has to be dealt with in a lot of languages. 🙂